Draft/RFC0001-PowerShell-Based-LCM.md

```diff
@@ -0,0 +1,92 @@
+---
+RFC: 0001
+Author: Ryan Bartram
+Status: Draft
+Version: 1.0
+Area: Engine
+Comments Due: 14/11/2019
+Plan to implement: <Yes | No>
+---
+
+# PowerShell Based LCM
+
+The Local Configuration Manager (LCM) was a windows component made available with the
+release of the WMF 4. It's the engine behind Desired State Configuration (DSC) and is used
+predominently as a Windows based configuration management tool.
+
+The current implementation is based on WMI/OMI and receives MOF files either from a Pull Server
+or by use of the cmdlets `Start-DscConfiguration` or `Publish-DscConfiguration`.
+
+The LCM can be configured to pull its configuration on a recurring basis and implements security
+by verifying signatures of MOFs or modules before they are downloaded.
+
+The LCM has 3 application modes `ApplyOnly`, `ApplyAndMonitor` and `ApplyAndAutoCorrect`. When
+the configuration is applied, it is applied in a serial fashion i.e. one resource after another.
+There is a dependency model defined within the configuration which enforces the order with which
+resources are deployed.
+
+## Motivation
+
+As a user, I want to replace the existing WMI/OMI based LCM on Windows and Linux with cross
+platform version that can perform all of the tasks of the existing LCM
+plus a lot lot more. So I can utilise my existing DSC knowledge to manage infrastructure
+all over my organisation. Onprem and in the cloud.
+
+I see this being the new Terraform, Polumi etc
+
+## Specification
+
+### User Experience
+
+```PowerShell
+# Command to generate new DSC MOF file
+Build-DscConfiguration
+
+# Command to execute DSC on a machine
+Invoke-DscConfiguration
+```
+
+### The basic Set
+This RFC proposes a way of using PowerShell to perform the task of the LCM:
+
+```PowerShell
+param([Resources[]]$resources)
+
+# Loop each resource and use the Invoke-DscResource cmdlet available in both Windows Powershell
+# and PowerShell Core to start each resource.
+
+foreach($resource in $resources) {
+    $InvokeParameters = @{
+        name = $resource.name
+        module = $resource.module
+        properties = $resource.properties
+    }
+
+    if ($false -eq (Invoke-DscResource @InvokeParameters -Method Test)) {
+        Invoke-DscResource @InvokeParameters -Method Set
+    }
+}
+```
+
+This is a very trivial implementation that just demonstrates how we could implement the set
+functionality of the LCM within PowerShell.
+
+### Dependency model
+
+I recommend we keep the near enough same data model for resources i.e. `resourceid`, `modulename`,
+`moduleversion` and `dependson`. I would make a seperate `properties` object with all the resource
+properties just to seperate the instance data from the meta data.
+
+On initial run, all the resources without any dependencies would get added to a "deployment" queue
```

Comment:
I'm not 100% sure I understand what you describe.
Maybe we'll have call to clarify.
Only remark here is that we should support/keep a default execution model similar to what DSC in Windows PS 5.1 has, so that it's easier for users to transition to it.

Reply:
it was just my way of describing how I would try to overcome the serial nature by which DSC executes today. I will try to reword, but we can also have a call to describe